### PR TITLE
AP_Notify: Support for OLED display by Alexey Kozin

### PR DESCRIPTION
**Pull request by dipspb**
https://github.com/ArduPilot/ardupilot/pull/5082#pullrequestreview-6356583

It seems github refused to accept rebased commits that were included in https://github.com/ArduPilot/ardupilot/pull/5082
So I have to create new branch and submit this new PR to provide all commits squashed to a single one with "git rebase -i HEAD~18".

### CODE REVIEW

**Comment by magicrub** [comment]
The github app is junk, I recommend sourcetree. What you need to do is force push study the rebase -i

**Comment by dipspb** [comment]
@magicrub Thank you, Tom. I didn't mean github app. I mean github site. When I pushed rebased commits to the same branch it doesn't complain anything. But it seems nothing has been changed on side of github repo. It still have all old commits as not rebased. Then I attempted to repeat push with force option and still have no luck. Then I created new branch.

**Comment by dipspb** [comment]
Minor fix added as separate commit https://github.com/ArduPilot/ardupilot/pull/5135/commits/e289523d3fbbd3d8e8e6539966d96cd38f28e89c into this PR because we have just discovered issue in hardware test and fixed it.

**Comment by magicrub** [comment]
@dipspb your changes were in effect on the other PR

**Comment by dipspb** [comment]
@magicrub Not quite sure I've got your point. What do you mean?

**Comment by magicrub** [comment]
@dipspb I mean the updates to the branch are on that PR. Everything was fine. Changing to a new PR was unnecessary. However, since you're making a new branch and a new PR it would have been nice to rebase onto master while you were at it but that's OK, there's no merge conflicts so it happens auto-magically

**Comment by magicrub** [comment]
I hate to be the one bringing the bad news, but you've squashed your commits too much. In a PR like this we'd really like to see the commits done to the vehicle be in a separate commit.

**Comment by magicrub** [comment]
If you aren't able to do that then maybe the one merging it can

**Comment by dipspb** [comment]
@magicrub I remastered my commits in accordance with recommendations from @mirkix https://github.com/ArduPilot/ardupilot/pull/5082#issuecomment-257025025 and following "submitting patches" document http://ardupilot.org/dev/docs/submitting-patches-back-to-master.html

And I have not a minor advice on what commits should be left separate. Nobody warned me and so on..

So my three questions at this moment:
1) How would I guess it?
2) Where I can read plain and clear document the covers issue you have just mentioned?
3) What I need to do?

Thank you very much in advance

p.s. The person who is merging, it's me.

**Comment by OXINARF** [comment]
@dipspb It is written there:

"Commits should be small, and do just one thing. If a change touches multiple libraries then there should be a separate commit per library, and a separate commit per vehicle directory. This is true even if it means that intermediate commits break the build."

Along with:

> Commit messages should be of the form:
> 
> ```
> Subsystem: brief description
> 
> Longer description...
> ```
> 

Anyway, there are previous comments I made in the previous PR that still haven't been addressed. Also, it has been found by Tridge that this driver (along with others) are doing I2C transactions in the main thread and although that looks to be fine in Linux it is a no-no for Pixhawk. That needs to be fixed before this can go in.

**Comment by dipspb** [comment]
@OXINARF thank you. Will redo. So one more branch/PR?

> this driver (along with others) are doing I2C transactions in the main thread and although that looks to be fine in Linux it is a no-no for Pixhawk. That needs to be fixed before this can go in.

How and where it need to be fixed? What about other drivers?

**Comment by dipspb** [comment]
Ok. I can see 4 (sort of) unresolved issues to do:
1) One missed usage of sprintf. To be done.
2) "ARMED message only when armed" https://github.com/ArduPilot/ardupilot/pull/5082#discussion_r85651181
3) I2C issue(s) https://github.com/ArduPilot/ardupilot/pull/5082#discussion_r85651314
4) Separate squashed commits for: AP_Notify, ArduCopter and ArduPlane

Right? I can see what to do with 1 and 4. And I need to see what to do with 2 and 3.

**Comment by OXINARF** [comment]
@dipspb Several points:

- when I said there are comments from me that weren't addressed I was referring to the first two points of https://github.com/ArduPilot/ardupilot/pull/5082#issuecomment-256953908
- regarding the I2C fixes you only need to worry about this driver since we can add it to Pixhawk without fixing it. You can see at https://github.com/ArduPilot/ardupilot/pull/5109/commits/91cc8b7f1ea4d4583aa6de6b774bb7a90800f9ef how Tridge fixed it. We also need to understand why @kozinalexey was seeing artifacts with the big I2C transfer (your point number 3).
- regarding the ARMED I just wanted to understand if that was the desired functionality, I haven't answered as I think @kozinalexey gave a good answer
- you don't need to do another PR, you need to split the commit and force push. To do that you will either need to use the git command line or use a GUI to work with Git (Tom recommended SourceTree and I can recommend it too). If you need help with this we can help in Gitter.

**Comment by magicrub** [comment]
>p.s. The person who is merging, it's me.

I meant the person who's merging into Ardupilot. We can do it

File: libraries/AP_Notify/AP_Notify.h [hunk 1]
```diff
@@ -110,11 +109,23 @@ class AP_Notify
...
     bool buzzer_enabled() const { return _buzzer_enable; }
+
+    static void send_text(const char *str) { snprintf(_send_text, sizeof(_send_text), "%s", str); }
+    static char* get_text() { return _send_text; }
+
 private:
     static NotifyDevice* _devices[];
-
+    static float _voltage;
+    static uint8_t _control_mode;
```
Comment (magicrub) [inline libraries/AP_Notify/AP_Notify.h]: why are _voltage and _control_mode static?

**Comment by dipspb** [comment]
Commits were reworked.

**Comment by dipspb** [comment]
Extra methods and fields like _voltage and _control_mode were fixed, they are moved to AP_Notify::flags.

**Comment by rmackay9** [comment]
I've had a quick look at this, the duplication of the flight mode names doesn't make me too happy.  It's something we would need to remember to update each time we add a new flight mode.  Not the end of the world but not great.  We already have a print_flight_mode function in each vehicle class although it sends to a "BetterStream" and it would probably be ugly to pass this into Notify.

I was actually surprised that this was written as a Notify object.   I was expecting it to be more like the FrSky protocol or a telemetry link to the ground station.  It would be nice for example to have it display the messages that normally appear on the mission planner HUD.. like the pre-arm check failures.

**Comment by dipspb** [comment]
Hello, @rmackay9 Unsure I've got your point about Notification vs Telemetry. Could you pls to clarify in more details? What's wrong with Notify object? We didn't do that, it was already implemented this way by @mirkix .

That's for mode names, IMHO it shouldn't be a big deal to refactor print_flight_mode by extracting get_flight_mode from it. Any other ideas are appreciated.

The only thing that frustrating me really much is a fact that review feedback notes are arriving one-by-one with a few days gap. May I have review notes for this PR all at once so I would able to fix all of them and then to be sure PR will be accepted?

**Comment by rmackay9** [comment]
The ability to display the messages is great.  I'll take responsibility for reviewing this PR in more detail and getting it in.

As a side note, I've got four screens coming (two of each of these):
https://www.adafruit.com/product/326
https://www.adafruit.com/product/938

File: libraries/AP_Notify/Display.cpp [hunk 2]
```diff
@@ -312,46 +387,40 @@ void Display::update()
...
-        update_prearm();
-        _flags.pre_arm_check = AP_Notify::flags.pre_arm_check;
-    }
-
-    if (_flags.gps_status != AP_Notify::flags.gps_status) {
-        update_gps();
-        _flags.gps_status = AP_Notify::flags.gps_status;
-    }
+	if (AP_Notify::flags.armed)
+		{
```
Comment (magicrub) [inline libraries/AP_Notify/Display.cpp]: tabs

**Comment by rmackay9** [comment]
My screens have arrived the 1.3" one at least is not immediately working when connected to I2C.
As far as I can tell there aren't any parameters that need to be set to enable it.
Hopefully I'll figure out what's wrong.

**Comment by magicrub** [comment]
Doh!

**Comment by kozinalexey** [comment]
1.3 inch OLED have another IC driver, dipspb always prepare code for 1.3 inch display type support, but he waiting for this pr will be accepted, else oled code integration process stops in infinite loop

File: libraries/AP_Notify/Display.cpp [hunk 3]
```diff
@@ -279,9 +280,82 @@ static const uint8_t _font[] = {
...
+    "----", //14 no mode for plane
+    "GUID", //  = 15,
+    "INIT", //  = 16,
+    "QSTB", //  = 17,
+    "QHVR", //  = 18,
+    "QLIT", //  = 19,
+    "QLND", //  = 20,
+    "QRTL" //   = 21
+};
+#else  //rover
```
Comment (night-ghost) [inline libraries/AP_Notify/Display.cpp]: I think that you forget AntennaTracker, where display with coordinates of drone is a long awaited feature.

Comment (night-ghost) [inline libraries/AP_Notify/Display.cpp]: That solution is very good but requires a some additional piece of hardware. But small display which shows Tracker mode for some time and then shows UAV coords will be a very good thing too.

Comment (dipspb) [inline libraries/AP_Notify/Display.cpp]: @night-ghost Thanks for your feedback, we will try to figure out how to implement it and probably to put it into separate PR

Comment (kozinalexey) [inline libraries/AP_Notify/Display.cpp]: idea: based minim-osd code  mavlink tft display device

Comment (magicrub) [inline libraries/AP_Notify/Display.cpp]: @dipspb I think we can live without Tracker support for now in an effort to get this PR merged in soon. We can add tracker support after this is in master, However, the bigger problem is the one @rmackay9 is having of it simply not working at all on the hardware he got to test this with. Please work with him on that ASAP.

Comment (dipspb) [inline libraries/AP_Notify/Display.cpp]: @magicrub I have it already discussed with @tridge in gitter. His OLED modules have SH1106 chip which is not compatible with SSD1306 used in this driver. I have support for SH1106 already implemented but OLED type auto-detection still incomplete. It is not kind of fix for something broken. Both parts, SH1106 and autodetection are definitely extensions of this feature.

Comment (dipspb) [inline libraries/AP_Notify/Display.cpp]: Anyway I going to add SH1106 and autodetection into this PR

Comment (OXINARF) [inline libraries/AP_Notify/Display.cpp]: @dipspb I posted in Gitter: the links @rmackay9 posted talk about SSD1306, where have you seen the SH1106?

Comment (dipspb) [inline libraries/AP_Notify/Display.cpp]: It doesn't matter too much, because such kind of modules frequently referred as being SSD1306 but indeed they SH1106. It's just because chips are pretty similar and even have shared some commands.

I know it because I have both types in my hands. 0.96" and 1.3".  The 1.3" module didn't work until I implemented support for SH1105.

BTW, seller may mistakenly consider it as SSD2306 because chip implemented as COG (chip on glass) with no clearly identified label outside.

Comment (dipspb) [inline libraries/AP_Notify/Display.cpp]: @rmackay9 Still have no information about FMC type that you have used to test OLED display. I need this to fix your issue with OLED. I guess wrong number of I2C bus was used here for some reason https://github.com/ArduPilot/ardupilot/blob/81180ca97b110f5c63e102c927dad189ef4fbd65/libraries/AP_Notify/Display_SSD1306_I2C.cpp#L22-L26

Comment (mariansoban) [inline libraries/AP_Notify/Display.cpp]: Hi, I've tried to test this OLED:
https://www.aliexpress.com/item/Wholesale-0-96-inch-4pin-White-OLED-Module-SSD1306-Drive-IC-128-64-I2C-IIC-Communication/32658908775.html
on my Pixhawk Lite yesterday by porting of changes above to AC 3.4.3, but I'm only getting "random" white dots instead of text on OLED. 
Any idea, what could be the issue?

Comment (mariansoban) [inline libraries/AP_Notify/Display.cpp]: just the compass and OLED. I'll test it in the evening, thanks for now.

Comment (Micheal-Kelly) [inline libraries/AP_Notify/Display.cpp]: "I'm only getting "random" white dots instead of text on OLED."
Like this?
img_0351

Try Alexey's pre-compiled example to test your hardware:
https://www.dropbox.com/s/djpkvc7d6tc58b9/ArduCopter-34rc1-pixhawk-v2-ssd1306test.zip?dl=0

Comment (mariansoban) [inline libraries/AP_Notify/Display.cpp]: "Like this?" - yep :)

Comment (kozinalexey) [inline libraries/AP_Notify/Display.cpp]: i tested both single colour ssd1306 modules and two-colour 
all work fine
may be you have defect board?

Comment (kozinalexey) [inline libraries/AP_Notify/Display.cpp]: two_color_oled

Comment (Micheal-Kelly) [inline libraries/AP_Notify/Display.cpp]: When I tried to modify Mirko's original BBBmini code to use with Pixhawk, instead of BBBmini,  I got the result above with random dots. I missed something that you did correctly Alexey. When I used your test code with the same SSD_1306 module and same fc it works fine. So I don't believe the module is defective.

Comment (mariansoban) [inline libraries/AP_Notify/Display.cpp]: OK, in the end I've made it work somehow.
This is the commit for Copter-3.4.3 that contains (hopefully all) the code changes from above:
https://github.com/mariansoban/ardupilot/commit/fb5ee4d4c2ddb5fc92379ab02fe085292e2655e4
But it doens't work:
dsc_0112

And this is the fix:
https://github.com/mariansoban/ardupilot/commit/2c7628f3077bfdcd454edfac05e80663223eadad

dsc_0113
It's Sergey's older version of SSD1306_I2C driver I guess, I've just added there method clear_screen() to fit library requirements.
Full Copter-3.4.3 with OLED support source code is here:
https://github.com/mariansoban/ardupilot/commits/Copter-3.4-oled

**Comment by jinchengde** [comment]
That's so great!

**Comment by dipspb** [comment]
I've spent a lot of time in attempt to implement OLED type auto-detection, but there is no success. It seems there is no reliable way to auto-detect OLED type. So I going to implement a single parameter like OLED_CONFIG that will contain both values: the i2c bus number to use and the type of OLED.  Will that do?

**Comment by magicrub** [comment]
Prefer sperate Params: _type and _addr.

**Comment by dipspb** [comment]
@magicrub agree, but there are 3 params to config indeed:

1) i2c bus number
2) i2c address (fixed value for most OLEDs, but options are possible)
3) OLED type

Couldn't it be too much? In case nobody have objections will go this way.

**Comment by magicrub** [comment]
Do we really need bus number?

**Comment by OXINARF** [comment]
I would start just with OLED type for now. If there's a real need to add other parameters in the future we can do it - like you said I2C address will be mostly fixed for an OLED driver and bus number shouldn't very important as having a fixed one is probably enough for most of our boards.

**Comment by dipspb** [comment]
Done, the NTF_DISPLAY_TYPE parameter introduced. Values are:

0 = Disabled (default)
1 = SSD1306
2 = SH1106

**Comment by dipspb** [comment]
tabs issue resolver

File: libraries/AP_Notify/AP_Notify.h [hunk 4]
```diff
@@ -111,10 +122,17 @@ class AP_Notify
     static const struct AP_Param::GroupInfo var_info[];
 
     bool buzzer_enabled() const { return _buzzer_enable; }
+
+    static void send_text(const char *str) { snprintf(_send_text, sizeof(_send_text), "%s", str); }
```
Comment (magicrub) [inline libraries/AP_Notify/AP_Notify.h]: calls to send_text() should probably do nothing unless this feature is enabled

Comment (dipspb) [inline libraries/AP_Notify/AP_Notify.h]: The send_text() is a static method so it unable access _display_type parameter variable which is instance one. Unsure I can make _display_type as a static. Also send_text() just prints string to the buffer. It is not a problem, because in case of _display_type == DISPLAY_OFF the buffer will not used to send data to display.

**Comment by dipspb** [comment]
Minor buffer initialization fix added https://github.com/ArduPilot/ardupilot/pull/5135/commits/593654ee7e616cfd55ac8d84ec2138cb64a7d3b2#diff-538a73bac9a5fbabe5668d4e34f232caR163

File: libraries/AP_Notify/AP_Notify.h (hunk 4 above)
Comment (magicrub) [inline libraries/AP_Notify/AP_Notify.h]: >Also send_text() just prints string to the buffer.

I agree that snprintf is the safest, but it *extremely* slow. It's not a simple mem copy, it's a string parsing operation. Even though it's a relatively simple one it's still much more complicated than a simple strncpy and I think it's safe to say that we will always have a null terminated string arrive here. Even not, you're clipping to the _send_text size so it's safe. I propose either using strcpy() and/or check if _type != DISPLAY_OFF.

See the comparison here: http://stackoverflow.com/questions/12275381/strncpy-vs-sprintf

Execution time, including parameter shuffling on/off call stack:

strcpy   43 instructions
sprintf  467 instructions
Program/ROM space allocated:

strcpy   56 bytes
sprintf  1488 bytes
RAM/stack space allocated:

strcpy   0 bytes
sprintf  15 bytes

Comment (dipspb) [inline libraries/AP_Notify/AP_Notify.h]: replaced to strncpy

**Comment by dipspb** [comment]
snprintf(..., "%s", ...) changed to strncpy

**Comment by magicrub** [comment]
Thanks @dipspb . I just did a check for snprintf and vsnprintf and it's used almost 70 times where strncpy is only used about 25 times. Maybe I gave you wrong advice?

Can anyone else chime in on this to sanity check me?

**Comment by dipspb** [comment]
@magicrub I pretty sure it was really good advice for that particular method. Thank you!

**Comment by dipspb** [comment]
Hi @rmackay9 , do you have any success testing your OLED modules with my final commits?

**Comment by rmackay9** [comment]
I've received a display from @dipspb and it works!  I don't understand why my existing displays from adafruit don't work.  Is the aliexpress link from above the best source of displays?

I've rebased on master and resolved a few conflicts. https://github.com/rmackay9/rmackay9-ardupilot/commits/oled-display2.

There are a few potential issues that I hope to make more progress on tomorrow:

- the front-end/back-end split is a slightly different method than we currently use (see other libraries like AP_Buzzer to see how we normally do it) so I will like re-format it to look more like other libraries.
- @tridge is concerned about the I2C traffic especially when flying so I will test this.
- @OXINARF (and perhaps others) will be unhappy about the vehicle specific #defines in the library to provide the flight modes.

**Comment by kozinalexey** [comment]
@rmackay9 oled display have i2c address solder jumper. maybe adafruit model have wrong default setting

**Comment by magicrub** [comment]
To address the excess i2c traffic during flight, can we disable this feature while flying and/or armed to get this PR in and then fix that later?

**Comment by kozinalexey** [comment]
display do not have traffic in armed state, after  ARM it clear screen , show message "ARMED" and stop any updates

**Comment by OXINARF** [comment]
> @OXINARF (and perhaps others) will be unhappy about the vehicle specific #defines in the library to provide the flight modes.

I couldn't think of a better option, so I'm not against it. It's not great to have it, but until someone does an AP_Telemetry library (which I envision as having subclasses in every vehicle) I think we'll continue to have these defines.

**Comment by rmackay9** [comment]
@OXINARF, Thanks.  As a side note, I've also been thinking that we need an AP_Telemetry library which would absorb our GCS_MAVLink library.

**Comment by lucasdemarchi** [comment]
> I couldn't think of a better option, so I'm not against it. It's not great to have it, but until someone does an AP_Telemetry library (which I envision as having subclasses in every vehicle) I think we'll continue to have these defines.

A method inside each vehicle that returns the flight mode would suffice and avoid the ifdef. It would be very simple while a more elaborate solution is not done.

**Comment by OXINARF** [comment]
I thought of that but I'm not a fan of vehicles calling libraries calling vehicles. Also, it wouldn't be a direct connection in this case because the Display object is created in AP_Notify (and that is what is created in the vehicle), which makes it even worse for me.

**Comment by lucasdemarchi** [comment]
Still better than a ifdef.

AP_Vehicle::get_flight_mode(), to be implemented by each vehicle. No reference passing, straight plain old boring function :)

**Comment by OXINARF** [comment]
I consider them almost equally bad that I don't care about the ifdef.

**Comment by rmackay9** [comment]
I'm going to close this PR because we have an updated PR now: https://github.com/ArduPilot/ardupilot/pull/5581
